### ISSUE TITLE
Kerberos authentication fails when accessing a DFS share on Linux

### ISSUE THREAD

**Issue by samuelwcm** [jborean93/smbprotocol#59]
Hi, I am encountering a `gssapi.raw.exceptions.BadNameError` when I try to access a DFS fileshare using Kerberos auth on Linux. I do not encounter the same error on Windows.

I suspect this is due to the DFS referral not quite working properly on Linux, perhaps it is generating the incorrect SPN? Possibly a similar issue to this: https://lists.samba.org/archive/samba-technical/2017-January/118057.html

I am still digging into the code to figure out what's going on here. I'll report back if I find any more clues.

```
>>> import smbclient as smb
>>> smb.listdir(r"\\fileserver\share$\dir")
['subdir1', 'subdir2']
>>> smb.listdir(r"\\company.com\dfs\share\dir")
Traceback (most recent call last):
  File "/opt/python37/lib/python3.7/site-packages/spnego/_context.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/opt/python37/lib/python3.7/site-packages/spnego/gss.py", line 435, in step
    out_token = self._context.step(in_token)
  File "<decorator-gen-15>", line 2, in step
  File "/opt/python37/lib/python3.7/site-packages/gssapi/_utils.py", line 155, in check_last_err
    return func(self, *args, **kwargs)
  File "<decorator-gen-5>", line 2, in step
  File "/opt/python37/lib/python3.7/site-packages/gssapi/_utils.py", line 128, in catch_and_return_token
    return func(self, *args, **kwargs)
  File "/opt/python37/lib/python3.7/site-packages/gssapi/sec_contexts.py", line 519, in step
    return self._initiator_step(token=token)
  File "/opt/python37/lib/python3.7/site-packages/gssapi/sec_contexts.py", line 540, in _initiator_step
    token)
  File "gssapi/raw/sec_contexts.pyx", line 245, in gssapi.raw.sec_contexts.init_sec_context
gssapi.raw.exceptions.BadNameError: Major (131072): An invalid name was supplied, Minor (100001): Success
 
The above exception was the direct cause of the following exception:
 
Traceback (most recent call last):
  File "/opt/python37/lib/python3.7/site-packages/smbprotocol/session.py", line 267, in connect
    out_token = context.step(in_token)
  File "/opt/python37/lib/python3.7/site-packages/spnego/negotiate.py", line 124, in step
    mech_token_out = self._step_spnego_token(in_token=mech_token_in)
  File "/opt/python37/lib/python3.7/site-packages/spnego/negotiate.py", line 220, in _step_spnego_token
    out_token = context.step(in_token=in_token)
  File "/opt/python37/lib/python3.7/site-packages/spnego/_context.py", line 91, in wrapper
    reraise(SpnegoError(base_error=native_err, context_msg=context))
  File "/opt/python37/lib/python3.7/site-packages/spnego/_compat.py", line 60, in reraise
    raise exc
spnego.exceptions.SpnegoError: SpnegoError (4294967295): Major (131072): An invalid name was supplied, Minor (100001): Success, Context: Processing security token
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/python37/lib/python3.7/site-packages/smbclient/_os.py", line 232, in listdir
    with SMBDirectoryIO(path, mode='r', share_access='r', **kwargs) as dir_fd:
  File "/opt/python37/lib/python3.7/site-packages/smbclient/_io.py", line 265, in __init__
    tree, fd_path = get_smb_tree(path, **kwargs)
  File "/opt/python37/lib/python3.7/site-packages/smbclient/_pool.py", line 287, in get_smb_tree
    connection_timeout=connection_timeout, connection_cache=connection_cache)
  File "/opt/python37/lib/python3.7/site-packages/smbclient/_pool.py", line 348, in register_session
    session.connect()
  File "/opt/python37/lib/python3.7/site-packages/smbprotocol/session.py", line 269, in connect
    raise SMBAuthenticationError("Failed to authenticate with server: %s" % str(err.message))
smbprotocol.exceptions.SMBAuthenticationError: Failed to authenticate with server: SpnegoError (4294967295): Major (131072): An invalid name was supplied, Minor (100001): Success, Context: Processing security token
```

**Comment by jborean93** [comment]
Yea I remember this being an issue for me where the DFS referral being returned to the client. On Windows it's typically able to resolve netbios hostnames like `server` to the fully qualified name `server.domain.com` so the SPN lookup works. It also knows more information about the domain and usually has a valid domain referral cache in place which just isn't possible to have by default on Linux.

Here is what I think is happening:

* You are attempting to connect to `\\company.com\dfs\share\dir`
* The connection pool looks into the referral cache to see if that path or any parent paths exist in the cache
  * Because this is the first time you are connecting to that path there won't be any referral
* Based on your code I don't see any explicit domain config settings so there won't be any domain referrals happening
* The client then connects to `company.com`, whatever that name resolves to
* Once connected to that server it will attempt to authenticate itself
  * Kerberos uses the hostname as part of the target name, in this case it will be `cifs@company.com`
  * In this case this SPN isn't actually valid for any principals on your domain as it's just the domain name itself

Because the SPN isn't valid then Kerberos authentication will fail with the error that you have. The "solution" that is currently in place for this setup is to pre-populate your domain information in the `smbclient.ClientConfig(domain_controller='dc01.company.com')`.

DFS is a complicated beast as there are multiple hosts that the client may need to talk to so that it has all the referral information. There are 3 types of requests that are used in DFS and each of them are cached by `smbclient`. These 3 types are

* [Domain Referral](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dfsc/68cf1aab-58e5-466e-bd8b-9f9a81a54b24) - Sent to the DC IPC$ share that gets a list of known domain names like `company.com` and `COMPANY`
* [DC Referral](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dfsc/f22bd18d-288c-45e7-983d-6e54868c0d29) - Sent to the DC IPC$ share to get a list of domain controllers available with a priority order
* [Root/Link Referral](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dfsc/f4bb505c-e81a-4110-a093-02871247a1c6) - Sent to the DC hint (or fallback if the hint fails) from the DC Referral response, contains the DFS servers that host the DFS roots

By default `smbclient` only uses the root/link referrals as unlike Windows we cannot know the domain information of a Linux client. By explicitly setting a domain controller in the config you enable the domain and DC referrals which should solve this issue. The connection logic will now look like:

* ClientConfig is set up with an explicit DC
  * Will connect to `\\{domain_controller}\IPC$`
  * Send a DFS domain referral to this IPC tree which returns the valid domain names for that DC, i.e. `company.com` and `COMPANY`
* You connect to `\\company.com\dfs\share\dir`
* First time so no root referral cache entry for this path
* The first component of your path is checked in the domain cache, in this case it should match `company.com` with the cache entry
* If there is a match it will reconnect to `\\dc01.company.com\IPC$` and send a DC referral request
  * The DC referral request returns a list of domain controllers in your environment with a hint as to what one is best to use
  * The DC referral is cached so this step only runs when the TTL has expired
* It will then connect to the `\\{DC referral hint}\IPC$` and send a root referral request for `\\company.com\dfs`
  * The root referral response will contain the root target, i.e. `\\dfs-server.company.com\dfs`
  * The root referral response is cached so further connections will use this information until the TTL is hit
* The path is rewritten and now the host is connecting to `\\dfs-server.company.com\dfs\share\dir`
* Now the SPN being used for this connection will be `cifs@dfs-server.company.com` which should be valid for the endpoint

There is 1 slight hiccup here where I've seen a root referral return the netbios name of the DFS server(s) (`dfs-server`) and not the DNS name (`dfs-server.company.com`). Typically on Windows it won't matter because it can resolve the netbios name to the actual Kerberos principal but Linux may not be able to do that. In this case your need to either

* [Configure DFS to use the fully qualified domain names](https://docs.microsoft.com/en-us/troubleshoot/windows-server/networking/configure-dfs-use-domain-names) in the referral responses
* Configure your GSSAPI implementation to work with unqualified names, you may even need to tweak DNS in some way as well

TLDR: Set `smbclient.ClientConfig(domain_controller='mydc.comapny.com')` so the client can gather the domain information required to build the proper hostnames when connecting to a DFS path.

Are you explicitly setting `auth='kerberos'` as per your changes at https://github.com/jborean93/smbprotocol/pull/58? The current behaviour should fallback to NTLM auth so I just wanted to check whether that is still happening and you are just explicitly requesting Kerberos only.

**Comment by jborean93** [comment]
> Might be worth adding a note about this requirement in #60?

Happy to document it wherever it is needed.

> Interestingly it wasn't falling back to NTLM though, it just immediately reraised the error thrown by gssapi.

That sounds like a bug that should be fixed in pyspnego. If Kerberos fails to generate the token then it should fall back to NTLM.

**Comment by jborean93** [comment]
Closing as the support for domain based DFS shares without knowing the domain controller is quite limited. Setting `smbclient.ClientConfig(domain_controller='...')` is highly recommended to ensure things like Kerberos continues to work and have `smbclient` properly resolve DFS paths before it connects rather than after a connection is made and fails.

If there are any remaining doc problems feel free to open a PR. If there are still authentication problems after setting the `domain_controller` config option then opening another issue with those details would be great.
